addon/-private/system/relationships/relationships-payloads.js

```diff
@@ -0,0 +1,446 @@
+      this._rhsPayloads = {};
+    } else {
+      // Edge case when we have a reflexive relationship to itself
+      //  eg user hasMany friends inverse friends
+      this._rhsPayloads = this._lhsPayloads;
+    }
+
+    // either canoical on push or pending & flush
+    this._pendingPayloads = [];
+  }
+
+  get(modelName, id, relationshipName) {
+    this._flushPending();
+
+    if (modelName === this._lhsModelName && relationshipName === this._lhsRelationshipName) {
+      return this._lhsPayloads[id];
+    } else {
+      return this._rhsPayloads[id];
+    }
+  }
+
+  push(modelName, id, relationshipName, relationshipData) {
+    this._pendingPayloads.push([modelName, id, relationshipName, relationshipData]);
+  }
+
+  unload(modelName, id, relationshipName) {
+    this._flushPending();
+
+    if (modelName === this._lhsModelName && relationshipName === this._lhsRelationshipName) {
+      this._unload(id, this._lhsPayloads, this._rhsPayloads);
+    } else {
+      this._unload(id, this._rhsPayloads, this._lhsPayloads);
+    }
+  }
+
+  _unload(id, payloads, inversePayloads) {
+    if (!this._inverseLoaded(payloads[id])) {
+      this._removeInverse(id, payloads[id], inversePayloads);
+      delete payloads[id];
+    }
+  }
+
+  _flushPending() {
+    let work = this._pendingPayloads.splice(0, this._pendingPayloads.length);
+    for (let i=0; i<work.length; ++i) {
+      let modelName = work[i][0];
+      let id = work[i][1];
+      let relationshipName = work[i][2];
+      let relationshipData = work[i][3];
+
+      let entry = {
+        payload: relationshipData,
+        inverse: false
+      };
+      // TODO: maybe delay this slightly?
+      let inverseEntry = {
+        payload: {
+          data: {
+            id: id,
+            type: modelName
+          }
+        },
+        inverse: true
+      }
+
+      if (modelName === this._lhsModelName && relationshipName === this._lhsRelationshipName) {
+        this._removeInverse(id, this._lhsPayloads[id], this._rhsPayloads);
+        this._lhsPayloads[id] = entry;
+        this._populateInverse(relationshipData, inverseEntry, this._rhsPayloads, this._rhsRelationshipIsMany);
+      } else {
+        this._removeInverse(id, this._rhsPayloads[id], this._lhsPayloads);
+        this._rhsPayloads[id] = entry;
+        this._populateInverse(relationshipData, inverseEntry, this._lhsPayloads, this._lhsRelationshipIsMany);
+      }
+    }
+  }
+
+  _inverseLoaded(entry) {
+    let data = entry && entry.payload && entry.payload.data;
+    if (!data) { return false; }
+
+    if (Array.isArray(data)) {
+      for (let i=0; i<data.length; ++i) {
+        if (this._hasRecordForId(this._store, data[i].type, data[i].id)) {
+          return true;
+        }
+      }
+      return false;
+    } else {
+      return this._hasRecordForId(this._store, data.type, data.id);
+    }
+  }
+
+  _hasRecordForId(store, type, id) {
```

Comment:
this method does not appear to rely on `this`, we should likely move it off the prototype and just to a module local helper function. Or, we should make it use `this._store` internally. 

When I see methods on `this` I feel it suggests it needs access to `this`.